pkg/capacityplanner/helpers.go

```diff
@@ -0,0 +1,347 @@
+		return err
+	}
+	_, acrToRemove := choseACinACRWithLessAC(ACMap{ac.Name: ac}, rh.acrMap, rh.acNameToACR)
+	if acrToRemove == nil {
+		logger.Infof("ACR holding AC %s not found. Skip deletion.", ac.Name)
+		return nil
+	}
+	if err := rh.removeACR(ctx, acrToRemove); err != nil {
+		return err
+	}
+	if ac == acReplacement {
+		return nil
+	}
+	if err := rh.replaceACInACR(ctx, acrToRemove.Name, ac, acReplacement); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (rh *ReservationHelper) update(ctx context.Context) error {
+	logger := util.AddCommonFields(ctx, rh.logger, "ReservationHelper.update")
+	acList, err := rh.capReader.ReadCapacity(ctx)
+	if err != nil {
+		logger.Errorf("failed to read AC list: %s", err.Error())
+		return err
+	}
+	acrList, err := rh.resReader.ReadReservations(ctx)
+	if err != nil {
+		logger.Errorf("failed to read ACR list: %s", err.Error())
+		return err
+	}
+	rh.acMap = buildACMap(acList)
+	rh.acrMap, rh.acNameToACR = buildACRMaps(acrList)
+
+	return nil
+}
+
+func (rh *ReservationHelper) removeACR(ctx context.Context, acr *acrcrd.AvailableCapacityReservation) error {
+	logger := util.AddCommonFields(ctx, rh.logger, "ReservationHelper.removeACR")
+	err := rh.client.Delete(ctx, acr)
+	if err == nil {
+		logger.Infof("ACR %s removed", acr.Name)
+		return nil
+	}
+	if k8serrors.IsNotFound(err) {
+		logger.Infof("ACR %s already removed", acr.Name)
+		return nil
+	}
+	logger.Errorf("Fail to remove ACR %s: %s", acr.Name, err.Error())
+	return err
+}
+
+func (rh *ReservationHelper) replaceACInACR(ctx context.Context, removedACR string,
+	ac, acReplacement *accrd.AvailableCapacity) error {
+	logger := util.AddCommonFields(ctx, rh.logger, "ReservationHelper.replaceACInACR")
+
+	acrToUpdate, ok := rh.acNameToACR[ac.Name]
+	if !ok {
+		logger.Infof("Can't find ACRs for AC %s", ac.Name)
+		return nil
+	}
+	for _, acrName := range acrToUpdate {
+		// ignore already removed ACR
+		if acrName == removedACR {
+			continue
+		}
+		acr, ok := rh.acrMap[acrName]
+		if !ok {
+			continue
+		}
+		replaced := false
+		for i := 0; i < len(acr.Spec.Reservations); i++ {
+			if acr.Spec.Reservations[i] == ac.Name {
+				acr.Spec.Reservations[i] = acReplacement.Name
+				replaced = true
+			}
+		}
+		if replaced {
+			err := rh.client.Update(ctx, acr)
+			if err != nil {
+				logger.Infof("Fail to update ACR %s: %s", acr.Name, err.Error())
+				return err
+			}
+			logger.Infof("ACR %s updated", acr.Name)
+		}
+	}
+	return nil
+}
+
+// NewReservationFilter returns new instance of ReservationFilter
+func NewReservationFilter(
+	acrFilter func(acr acrcrd.AvailableCapacityReservation) bool,
+	acFilter func(ac accrd.AvailableCapacity) bool) *ReservationFilter {
+	return &ReservationFilter{
+		acrFilter: acrFilter,
+		acFilter:  acFilter,
+	}
+}
+
+// ReservationFilter helper for working with ACR based reservations
+type ReservationFilter struct {
+	// additional filter for ACR
+	acrFilter func(acr acrcrd.AvailableCapacityReservation) bool
```

Comment:
I cannot see, where this functions `acrFilter` and `acFilter` are used in code, I see lines like `NewReservationFilter(nil, nil)`, so these functions are unnecessary, as I see. So could you please explain why do we need them?

Reply:
Agree. Users of this code where removed. Now this code is useless and should be removed.
Done